src/java/org/apache/cassandra/index/sai/plan/VectorTopKProcessor.java

```diff
@@ -187,6 +190,27 @@ public <U extends Unfiltered, R extends BaseRowIterator<U>, P extends BasePartit
                 for (var uf: pr.tombstones)
                     addUnfiltered(unfilteredByPartition, pr.partitionInfo, uf);
             }
+        } else if (partitions instanceof StorageAttachedIndexSearcher.ScoreOrderedResultRetriever) {
+            var retriever = (StorageAttachedIndexSearcher.ScoreOrderedResultRetriever) partitions;
+            // FilteredPartitions does not implement ParallelizablePartitionIterator.
+            // Realistically, this won't benefit from parallelizm as these are coming from in-memory/memtable data.
+            int rowsMatched = 0;
+            while (retriever.hasNext() && rowsMatched < limit)
+            {
+                // have to close to move to the next partition, otherwise hasNext() fails
+                try (var partitionRowIterator = retriever.next())
+                {
+                    var iter = (StorageAttachedIndexSearcher.ScoreOrderedResultRetriever.PrimaryKeyIterator) partitionRowIterator;
+                    PartitionResults pr = processPartition(iter, iter.scoredPrimaryKey, retriever.updatedKeys);
```

Comment:
@jbellis asked if we could get this merged and then add the parallelization as a follow up.

Reply:
why does this need to be a special branch?  can we fold it into the main logic by adjusting our interfaces to accommodate it?

Reply:
It gets a special branch right now because we need to get the `ScoredPrimaryKey` to determine if the row is an update or not. Also, we need a branch so we can skip post filtering of the result set. Once we get `LIMIT` valid rows, we return the result.

I wonder if we really need to implement parallelization. I left the branch in here for now, but we might be close to a place where the cost to context switch isn't worth it. The only time I can really imagine we'll want it is when we have a large individual query limit and very low latency is required.

We can make parallel reads work, but we'll need to track extra state in order to properly dedupe results while also handling updated rows.

@jbellis - do you think you'd be able to run a benchmark on this branch to see how we should prioritize parallelization? As it is, here is a screen shot of a low cardinality query with limit 3 on 1536 dimension vectors. We spend over 60% of our time in jvector while only 11.5% of the time is in VTKP.

![Screen Shot 2024-01-18 at 5 33 21 PM](https://github.com/datastax/cassandra/assets/47911938/c02a55c7-9536-4c18-a561-63c1783ebdef)

Here is the a profile of the same workload run on latest `vsearch`. We spend 29% of the time in VTKP with an additional 14% of the time run on the executor that parallelizes row materialization. 

![Screen Shot 2024-01-18 at 5 38 06 PM](https://github.com/datastax/cassandra/assets/47911938/f76fb2f8-0ed0-4adf-afc4-e5fa5fb56881)